keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,680 @@
+The behaviors are as follows:
+
+- Foreground cascading deletion:
+  - `blockOwnerDeletion=true`: Deletion of Deployment A isn't completed until Replicaset B is deleted
+  - otherwise: Deployment A is deleted immediately
+- Background cascading deletion: Deployment A is deleted immediately
+
+#### Namespace Deletion
+
+It may also not be clear how [namespace deletion](https://kubernetes.io/docs/tasks/administer-cluster/namespaces/#deleting-a-namespace) behaves with lien.
+Therefore, this section describes the behavior.
+
+In summary, lien will block the namespace deletion, but would not block deletion of any resources in the namespace individually.
+
+If a lien is set to a namespace, only a deletion request to the namespace is blocked.
+Users can still request to delete each object in the namespace directly, but they can't request to delete it through the namespace deletion.
+
+If a lien is set to one of objects in a namespace, only a deletion request to the object is blocked.
+Users can still request to delete its namespace or other objects in the namespace.
+When deletion of the namespace A is requested, the deletion isn't completed until the object with lien is deleted.
+
+#### Block Adding Additional `Liens` while Deleting
+
+`Liens` shouldn't be added after `DeletionTimestamp` is non-nil, which means pre-deletion processes for finalizers are being handled.
+Adding `Liens` while deleting itself won't do any harm for the cluster, because the resource will be deleted without any additional deletion requests that will be blocked by `Liens`.
+However, users may think that the successful addition of the `Liens` means that the resource won't be deleted until the `Liens` are deleted, which isn't true.
+To avoid such a misunderstanding from users, a request for adding additional `Liens` to a resource with non-nil `DeletionTimestamp` should be blocked by api-server.
+
+#### Format of `Liens`
+
+As a result of discussions, format of `Liens` is decided as a slice of strings and it needs to be used per controller(per user) basis.
+The main reasons are scalability and better aligned with `Finalizers`.
+
+- Per controller(Per user) `Liens`
+  - Pros:
+    - Aligned with `Finalizers`
+    - Good scalability (Size of the field and number of API calls required)
+  - Cons:
+    - Bad UX (Users won't know which other objects are actually using the object)
+    - Complex consuming controllers' implementation (Each controller needs to track the using objects)
+- Per reason(Per object) `Liens`
+  - Pros:
+    - Good UX (Users can know which other objects are actually using the object)
+    - Simple consuming controllers' implementation (Lien tracks the using objects)
+  - Cons:
+    - Not aligned with `Finalizers`
+    - Bad scalability (Size of the field and number of API calls required)
+
+#### Race of removing/adding liens
```

Comment:
> OTOH if you want a runtime referential integrity feature, then we have to permission the addition and removal of liens to make sure the wrong parties can't break the referential integrity. This sounds very complicated to me and I don't really want to do it if we don't have to.

@lavalamp - I'm somewhat on the fence here. I agree that "ability to ignore existing liens" is a very simple mitigation to the problem, and that we need to allow certain users to ignore liens. However, ideally it would be great if we could limit the exact set of users that can do that.

That said - thinking more about it, I guess it's solvable. I can create a ValidatingWebhook that via admission request gets the DeleteOptions (I'm assuming that the "NotBlockOnLien" (or however we call it) option will be added to DeleteOptions) and it can allow/reject the request based on UserInfo.
So it seems that technically it is possible to achieve what I need.
Or am I missing something above? [If not - I'm onboard with the currently proposed mitigation.]

@deads2k

Reply:
Applied @wojtek-t 's idea to the KEP.

Reply:
I hadn't thought about doing it in a webhook, that would make it optional functionality. We could get very similar functionality by e.g. adding an RBAC verb of "ignoreliens", and checking that verb when someone attempts to use the DeleteOption. I wonder if this would be acceptable to @deads2k? I think it's much simpler than controlling who is allowed to add the liens.

Reply:
Modifying RBAC would of course work too, but I had an impression that you're both against that.
If that's not the case - I think that would be a preferred option...

Reply:
Just for your information, I've implemented a prototype of force delete [here](https://github.com/mkimuram/kubernetes/commits/lienv2-forcedel) (Only the last 5 commits are for force delete). It only allows users in "system:masters" group to specify `NotBlockOnLien` delete option, which can be specified via `--not-block-on-lien=true` flag in kubectl delete command. I hope that this prototype helps further discussion.

Please also see the console log below:

<details>

- With a user in "system:masters" group
```
kubectl create secret generic test-secret --from-literal='username=my-app' --from-literal='password=39528$vdg7Jb'
kubectl patch secret test-secret -p '{"metadata":{"liens":["never.delete/me"]}}' --type=merge

cluster/kubectl.sh delete secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: deletion not allowed by liens

cluster/kubectl.sh delete --not-block-on-lien=true secret test-secret
secret "test-secret" deleted
```

- With a user __not__ in "system:masters" group (the test user, or mynamespace-user, is created by [this way](https://jeremievallee.com/2018/05/28/kubernetes-rbac-namespace-user.html))
```
kubectl create secret generic test-secret --from-literal='username=my-app' --from-literal='password=39528$vdg7Jb'
kubectl patch secret test-secret -p '{"metadata":{"liens":["never.delete/me"]}}' --type=merge

cluster/kubectl.sh delete secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: deletion not allowed by liens

cluster/kubectl.sh delete --not-block-on-lien=true secret test-secret
Error from server (Forbidden): secrets "test-secret" is forbidden: specifying NotBlockOnLien=true delete option is not allowd to user system:serviceaccount:mynamespace:mynamespace-user
```

</details>

Reply:
Added this to the KEP as alpha's scope. Note that kube-apiserver's option part of the current description isn't implemented in the prototype, so I'm not sure that it really works (Also, please check if setting `system:controller:namespace-controller` user to default will allow namespace lifecycle controller to force delete).